In [1]:
#Guided Project: Exploring Ebay Car Sales Data
##autos.csv

In [2]:
import pandas as pd
import numpy as np
autos=pd.read_csv('autos.csv',encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
data_col=list(autos.columns)

In [7]:
print(data_col)

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer', 'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode', 'lastSeen']


In [8]:
for index,row in enumerate(data_col):
    if row == 'yearOfRegistration':
        data_col[index]= 'registration_year'
        
    elif row == 'monthOfRegistration':
        data_col[index]= 'registration_month'
        
    elif row == 'notRepairedDamage':
        data_col[index]= 'unrepaired_damage'
        
    elif row == 'dateCreated':
        data_col[index]= 'ad_created'
    
    else:
        count=-1
        for char in row:
            count += 1
            if ord(char) in range(65,90):
                row=row[:count] + '_' + row[count:]
                count += 1
       
        row=row.lower()
        data_col[index]= row
        
print(data_col)
               
        

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


In [9]:
autos.columns=data_col

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Above i changed some column names & convert camel case to snake case

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

##### seller column has 49999 value = private among 50000. It should be dropped.
#### offer_type,abtest column are like seller,those have almost one value through the column
#### price,odometer column are numeric but stored as text

In [12]:
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

In [13]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [14]:
import re
autos['price']= autos['price'].str.replace('(\D)','')

In [15]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: object

In [16]:
autos['price']=autos['price'].astype(int)
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int32

In [17]:
autos['odometer_km']=autos['odometer_km'].str.replace('\D','')
autos['odometer_km']=autos['odometer_km'].astype(int)
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int32

In [18]:
autos.describe()

price  registration_year     power_p_s    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [19]:
autos['odometer_km'].unique().shape

(13,)

In [20]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [21]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

### all odometer value seem realistic. So we r not removing any data. 

In [22]:
autos['price'].unique().shape

(2357,)

In [23]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [24]:
autos['price'].value_counts().sort_index(ascending=True).tail(50)

98500       1
99000       2
99900       2
104900      1
105000      2
109999      1
114400      1
115000      1
115991      1
116000      1
119500      1
119900      1
120000      2
128000      1
129000      1
130000      1
135000      1
137999      1
139997      1
145000      1
151990      1
155000      1
163500      1
163991      1
169000      1
169999      1
175000      1
180000      1
190000      1
194000      1
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64

In [25]:
autos['price'].value_counts().sort_index(ascending=True).head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64

###  we can see that lowest price is 0(unrealistic) and highest price is 99999999  (which is unrealistic for an old car.)
### So we r removing the ad that is lower than 50 dollar and higher than 350000 .

In [26]:
bool=autos['price'] >350000
indexnames= autos.loc[bool].index
autos.drop(indexnames, inplace =True)

In [27]:
print(autos['price'].value_counts().sort_index(ascending=True).tail(50))
autos.describe()

84000     1
84997     1
85000     1
86500     1
88900     1
89000     1
89900     1
93000     2
93911     1
94999     1
98500     1
99000     2
99900     2
104900    1
105000    2
109999    1
114400    1
115000    1
115991    1
116000    1
119500    1
119900    1
120000    2
128000    1
129000    1
130000    1
135000    1
137999    1
139997    1
145000    1
151990    1
155000    1
163500    1
163991    1
169000    1
169999    1
175000    1
180000    1
190000    1
194000    1
197000    1
198000    1
220000    1
250000    1
259000    1
265000    1
295000    1
299000    1
345000    1
350000    1
Name: price, dtype: int64


price  registration_year     power_p_s    odometer_km  \
count   49986.000000       49986.000000  49986.000000   49986.000000   
mean     5721.525167        2005.075721    116.341196  125736.506222   
std      8983.617820         105.727161    209.218012   40038.133399   
min         0.000000        1000.000000      0.000000    5000.000000   
25%      1100.000000        1999.000000     70.000000  125000.000000   
50%      2950.000000        2003.000000    105.000000  150000.000000   
75%      7200.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        49986.000000         49986.0  49986.000000  
mean             5.723723             0.0  50812.804225  
std              3.711839             0.0  25777.404967  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49571.000000  
75%              9.000000             0.0  71522.000000  
max             12.000000             0.0  99998.000000

#### Above we dltd the rows where price is greater than 350000. 
#### Now number of rows turn to 49986 from 50000!

In [28]:
bool=autos['price'] <50
indexnames= autos.loc[bool].index
autos.drop(indexnames, inplace =True)
print(autos['price'].value_counts().sort_index(ascending=True).head(50))
autos.describe()

50      49
55       2
59       1
60       9
65       5
66       1
70      10
75       5
79       1
80      15
89       1
90       5
99      19
100    134
110      3
111      2
115      2
117      1
120     39
122      1
125      8
129      1
130     15
135      1
139      1
140      9
145      2
149      7
150    224
156      2
160      8
170      7
173      1
175     12
179      1
180     35
185      1
188      1
190     16
193      1
195      2
198      1
199     41
200    266
205      1
210      1
215      2
217      1
219      1
220     33
Name: price, dtype: int64


price  registration_year     power_p_s    odometer_km  \
count   48347.000000       48347.000000  48347.000000   48347.000000   
mean     5915.456099        2004.718845    117.471198  125875.752373   
std      9071.625794          87.786604    201.002968   39618.501091   
min        50.000000        1000.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     73.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7499.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        48347.000000         48347.0  48347.000000  
mean             5.795106             0.0  50995.058266  
std              3.679678             0.0  25742.517400  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30823.000000  
50%              6.000000             0.0  49733.000000  
75%              9.000000             0.0  71679.000000  
max             12.000000             0.0  99998.000000

#### Above we dltd the rows where price is lower than 50 dollar. 
#### Now number of rows turn to 48347!

In [30]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036031
2016-03-08    0.033260
2016-03-09    0.033053
2016-03-10    0.032225
2016-03-11    0.032556
2016-03-12    0.036983
2016-03-13    0.015658
2016-03-14    0.036672
2016-03-15    0.034294
2016-03-16    0.029495
2016-03-17    0.031584
2016-03-18    0.012907
2016-03-19    0.034749
2016-03-20    0.037789
2016-03-21    0.037293
2016-03-22    0.032949
2016-03-23    0.032267
2016-03-24    0.029433
2016-03-25    0.031501
2016-03-26    0.032267
2016-03-27    0.031088
2016-03-28    0.034894
2016-03-29    0.034108
2016-03-30    0.033715
2016-03-31    0.031832
2016-04-01    0.033735
2016-04-02    0.035535
2016-04-03    0.038575
2016-04-04    0.036528
2016-04-05    0.013051
2016-04-06    0.003165
2016-04-07    0.001386
Name: date_crawled, dtype: float64

#####  2016-04-03 ,maximum ad were first crawled. The percentage is 0.03868 

In [31]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038824
2016-04-04    0.036900
2016-04-05    0.011769
2016-04-06    0.003247
2016-04-07    0.001241
Name: ad_created, Length: 76, dtype: float64

In [32]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001076
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007343
2016-03-09    0.009639
2016-03-10    0.010611
2016-03-11    0.012390
2016-03-12    0.023745
2016-03-13    0.008873
2016-03-14    0.012638
2016-03-15    0.015885
2016-03-16    0.016464
2016-03-17    0.028089
2016-03-18    0.007322
2016-03-19    0.015802
2016-03-20    0.020622
2016-03-21    0.020601
2016-03-22    0.021408
2016-03-23    0.018595
2016-03-24    0.019774
2016-03-25    0.019174
2016-03-26    0.016775
2016-03-27    0.015575
2016-03-28    0.020849
2016-03-29    0.022318
2016-03-30    0.024738
2016-03-31    0.023828
2016-04-01    0.022876
2016-04-02    0.024883
2016-04-03    0.025172
2016-04-04    0.024510
2016-04-05    0.124972
2016-04-06    0.221627
2016-04-07    0.132087
Name: last_seen, dtype: float64

#### maximum ad was last seen by crawler at 2016-04-06. percentage is 0.22100

In [33]:
autos['registration_year'].describe()

count    48347.000000
mean      2004.718845
std         87.786604
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

##### min is 1000 which is inappropriate because car was not even invented then!
##### max is 9999 which means in future!

In [34]:
autos['registration_year']

0        2004
1        1997
2        2009
3        2007
4        2003
         ... 
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48347, dtype: int64

In [35]:
bool_2016=autos['registration_year'] > 2016
autos.loc[bool_2016,'registration_year'].value_counts().sort_index()

2017    1385
2018     468
2019       2
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

In [36]:
autos.loc[bool_2016,'registration_year'].describe()

count    1869.000000
mean     2050.937935
std       440.930277
min      2017.000000
25%      2017.000000
50%      2017.000000
75%      2018.000000
max      9999.000000
Name: registration_year, dtype: float64

#### number of year greater than 2016 is 1869

In [37]:
bool_1900=autos['registration_year'] < 1900
print(autos.loc[bool_1900,'registration_year'].value_counts().sort_index())
autos.loc[bool_1900,'registration_year'].describe()

1000    1
1001    1
1111    1
1800    2
Name: registration_year, dtype: int64


count       5.000000
mean     1342.400000
std       420.158661
min      1000.000000
25%      1001.000000
50%      1111.000000
75%      1800.000000
max      1800.000000
Name: registration_year, dtype: float64

#### number of year smaller than 1900 is 5

#### so,total number outside 1900-2016 is 1874!

#### As the number is much large, we should go for more custom logic.

In [38]:
autos['brand'].value_counts()

volkswagen        10293
bmw                5246
opel               5241
mercedes_benz      4633
audi               4150
ford               3371
renault            2317
peugeot            1426
fiat               1260
seat                915
skoda               777
nissan              740
mazda               739
smart               691
citroen             684
toyota              611
hyundai             482
sonstige_autos      460
volvo               435
mini                417
mitsubishi          395
honda               387
kia                 345
alfa_romeo          320
suzuki              285
porsche             282
chevrolet           274
chrysler            169
dacia               129
daihatsu            122
jeep                108
subaru              101
land_rover           99
saab                 79
daewoo               75
jaguar               74
trabant              67
rover                65
lancia               54
lada                 29
Name: brand, dtype: int64

In [39]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212898
bmw               0.108507
opel              0.108404
mercedes_benz     0.095828
audi              0.085838
ford              0.069725
renault           0.047924
peugeot           0.029495
fiat              0.026062
seat              0.018926
skoda             0.016071
nissan            0.015306
mazda             0.015285
smart             0.014293
citroen           0.014148
toyota            0.012638
hyundai           0.009970
sonstige_autos    0.009515
volvo             0.008997
mini              0.008625
mitsubishi        0.008170
honda             0.008005
kia               0.007136
alfa_romeo        0.006619
suzuki            0.005895
porsche           0.005833
chevrolet         0.005667
chrysler          0.003496
dacia             0.002668
daihatsu          0.002523
jeep              0.002234
subaru            0.002089
land_rover        0.002048
saab              0.001634
daewoo            0.001551
jaguar            0.001531
trabant           0.001386
r

In [40]:
booli= autos['brand'].value_counts(normalize=True) > 0.05

In [41]:
print(booli)

volkswagen         True
bmw                True
opel               True
mercedes_benz      True
audi               True
ford               True
renault           False
peugeot           False
fiat              False
seat              False
skoda             False
nissan            False
mazda             False
smart             False
citroen           False
toyota            False
hyundai           False
sonstige_autos    False
volvo             False
mini              False
mitsubishi        False
honda             False
kia               False
alfa_romeo        False
suzuki            False
porsche           False
chevrolet         False
chrysler          False
dacia             False
daihatsu          False
jeep              False
subaru            False
land_rover        False
saab              False
daewoo            False
jaguar            False
trabant           False
rover             False
lancia            False
lada              False
Name: brand, dtype: bool


we selected brands that have over a
certain percentage of the total values (e.g. > 5%). and those are 
volkswagen        0.212898
bmw               0.108507
opel              0.108404
mercedes_benz     0.095828
audi              0.085838
ford              0.069725

In [42]:
brand_mean_price={}
count=0
for name in booli:
    if name:
        brand_name=booli.index[count]
        all_row_of_that_brand_name= autos['brand'] == brand_name
        brand_mean_price[brand_name]=autos.loc[all_row_of_that_brand_name,'price'].mean()
    count += 1
print(brand_mean_price)

{'volkswagen': 5354.7044593413, 'bmw': 8305.433473122379, 'opel': 2964.7975577179927, 'mercedes_benz': 8571.013598100582, 'audi': 9252.86, 'ford': 3740.5440522100266}


In [44]:
brand_mean_mileage={}
count=0
for name in booli:
    if name:
        brand_name=booli.index[count]
        all_row_of_that_brand_name= autos['brand'] == brand_name
        brand_mean_mileage[brand_name]=autos.loc[all_row_of_that_brand_name,'odometer_km'].mean()
    count += 1
print(brand_mean_mileage)

{'volkswagen': 128980.86077917031, 'bmw': 132782.11971025544, 'opel': 129503.91146727724, 'mercedes_benz': 130928.12432549104, 'audi': 129613.2530120482, 'ford': 124378.52269356274}


In [46]:
bmp_series= pd.Series(brand_mean_price)
print(bmp_series)

volkswagen       5354.704459
bmw              8305.433473
opel             2964.797558
mercedes_benz    8571.013598
audi             9252.860000
ford             3740.544052
dtype: float64


In [47]:
bmm_series= pd.Series(brand_mean_mileage)
print(bmm_series)

volkswagen       128980.860779
bmw              132782.119710
opel             129503.911467
mercedes_benz    130928.124325
audi             129613.253012
ford             124378.522694
dtype: float64


In [48]:
df= pd.DataFrame(bmp_series,columns=['mean_price'])
df

mean_price
volkswagen     5354.704459
bmw            8305.433473
opel           2964.797558
mercedes_benz  8571.013598
audi           9252.860000
ford           3740.544052

In [53]:
df.loc[:,'mean_mileage']= bmm_series
df

mean_price   mean_mileage
volkswagen     5354.704459  128980.860779
bmw            8305.433473  132782.119710
opel           2964.797558  129503.911467
mercedes_benz  8571.013598  130928.124325
audi           9252.860000  129613.253012
ford           3740.544052  124378.522694

In [54]:
df.sort_index()

mean_price   mean_mileage
audi           9252.860000  129613.253012
bmw            8305.433473  132782.119710
ford           3740.544052  124378.522694
mercedes_benz  8571.013598  130928.124325
opel           2964.797558  129503.911467
volkswagen     5354.704459  128980.860779

if we sort according to price, the table looks like below:

audi	        9252.860000||||||||||||||129613.253012

mercedes_benz	8571.013598||||||||||||||130928.124325

bmw	            8305.433473||||||||||||||132782.119710

volkswagen	    5354.704459||||||||||||||128980.860779

ford	        3740.544052||||||||||||||124378.522694

opel	        2964.797558||||||||||||||129503.911467


We see that audi,mercedes_benz, bmw are most expensive in order. Expenses are inversely proportional to mileage.Less mileage has higher price.

Ford and Opel are lowest expensive in order and both of them are also inversely proportional to mileage.Though their mileage are less than first three,but they are cheaper than them!

Volkswagen is in between.It's mileage is lower than first three, it's still cheaper than them. Again mileage is greater than ford,but expensive than ford!


